### PR TITLE
Wip client auth revision

### PR DESCRIPTION
**Pull request by ekr**
This is still a WIP but is posted so that people can review.

### CODE REVIEW

File: draft-ietf-tls-tls13.md [hunk 1]
```diff
@@ -1532,55 +1532,68 @@ derivation scheme for all cipher modes.
...
-
-            {} Indicates messages protected using keys
-               derived from the ephemeral secret.
-
-            [] Indicates messages protected using keys
-               derived from the master secret.
+       Client                                               Server
+  
+Key  ^ ClientHello                                                  
+Exch v  + KeyShare               -------->
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: The ^ and v without a line joining them are a little hard to follow.  What about:

```
v---- Key exchange
 ClientHello
       + KeyShare              -------->
                                                       ServerHello
                                                      + KeyShare
v---- Server parameters
                                             {EncryptedExtensions}
                                          {ServerConfiguration*}
                                           {CertificateRequest*} 
v---- Authentication
                                                    {Certificate*}
                                            {CertificateVerify*}
                                <--------              {Finished}
     {Certificate*}
     {CertificateVerify*}
     {Finished}                -------->
```

Also, this doesn't allow the server to speak first, is that intentional?

File: draft-ietf-tls-tls13.md [hunk 2]
```diff
@@ -1601,39 +1610,34 @@ CertificateRequest
...
-  client is not authenticating via a certificate. [{{client-certificate-verify}}]
+: a signature over the entire handshake using the public key
+  in the Certificate message. This message will be omitted if the
+  server is not authenticating via a certificate. [{{certificate-verify}}]
 
 Finished
-: a MAC over the entire handshake computed using the Static Secret
-  and providing key confirmation. [{{server-finished}}]
+: a MAC over the entire handshake. This message provides key confirmation, binds the endpoint's identity
+  to the exchanged keys, and in some modes (see {{zero-rtt-exchange}})
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: Do you mean to reference the 0RTT handshake, or do you mean to refer to PSK/resumption?

File: draft-ietf-tls-tls13.md (hunk 1 above)
Comment (ekr) [inline draft-ietf-tls-tls13.md]: Why do you think it doesn't allow that?

File: draft-ietf-tls-tls13.md (hunk 2 above)
Comment (ekr) [inline draft-ietf-tls-tls13.md]: Both, actually.

File: draft-ietf-tls-tls13.md (hunk 1 above)
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: Because it shows the server's application data after the server has received the Finished.

File: draft-ietf-tls-tls13.md [hunk 3]
```diff
@@ -1703,42 +1707,62 @@ described below.
...
+               [] Indicates messages protected using keys
+                  derived from the master secret.
 ~~~
 {: #tls-0-rtt title="Message flow for a zero round trip handshake"}
 
+As shown in Figure {{tls-0-rtt}}, the Zero-RTT data is just added
+to the 1-RTT handshake in the first flight. Specifically, the
+client sends its Authentication messages after the ClientHello,
+followed by any application data. The rest of the handshake
+messages are the same as with Figure {{tls-full}}. This implies
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: `Figure {{figure-ref}}` puts the word "Figure" into the final text twice.

File: draft-ietf-tls-tls13.md [hunk 4]
```diff
@@ -1703,42 +1707,62 @@ described below.
...
+that the server can request client authentication even if the
+client offers a certificate on its first flight. This isn't
+necessarily useful but is consistent with the server being
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: I think that you might want to avoid statements like "this isn't necessarily useful"

File: draft-ietf-tls-tls13.md [hunk 5]
```diff
@@ -2887,6 +2791,8 @@ Structure of this message:
            CertificateExtension certificate_extensions<0..2^16-1>;
        } CertificateRequest;
 
+[[TODO: We agreed to add a context string here.]]
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: a context string, fresh entropy, or both?

File: draft-ietf-tls-tls13.md [hunk 6]
```diff
@@ -3025,17 +2931,253 @@ processing the server's Certificate, CertificateVerify, and
...
+
+The following table defines the Handshake Context and MAC Key
+for each scenario:
+
+~~~~
+Mode             Handshake Context                        Base Key
+----             -----------------                        --------
+0-RTT            ClientHello + ServerConfiguration        xSS
+                             + Server Certificate
+                             + CertificateRequest
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: I think that this needs to make it clear that these extras are cached, otherwise ServerConfiguration here can be confused for ServerConfiguration in the 1-RTT case.

File: draft-ietf-tls-tls13.md [hunk 7]
```diff
@@ -3025,17 +2931,253 @@ processing the server's Certificate, CertificateVerify, and
...
+Post-Handshake   ClientHello ... ClientFinished +         master_secret
+                 CertificateRequest
+~~~~
+Note 1: The ServerConfiguration, CertificateRequest, and Server Certificate
+in the 0-RTT case are the messages from the handshake where the
+ServerConfiguration was established.
+
+Note 2: The Handshake Context for the last three rows does not include
+any 0-RTT handshake messages, regardless of whether 0-RTT is used.
+
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: An extra note required to point out that ClientHello ... HelloRetryRequest ... ClientHello is included in all cases.

File: draft-ietf-tls-tls13.md [hunk 8]
```diff
@@ -3025,17 +2931,253 @@ processing the server's Certificate, CertificateVerify, and
...
+> The certificate MUST be appropriate for the negotiated cipher suite's key
+exchange algorithm and any negotiated extensions.
+
+Structure of this message:
+
+%%% Authentication Messages
+       opaque ASN1Cert<1..2^24-1>;
+
+       struct {
+           ASN1Cert certificate_list<0..2^24-1>;
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: TODO: add correlator/context id

File: draft-ietf-tls-tls13.md [hunk 9]
```diff
@@ -3025,17 +2931,253 @@ processing the server's Certificate, CertificateVerify, and
...
+
+All certificates provided by the server MUST be signed by a
+hash/signature algorithm pair that appears in the "signature_algorithms"
+extension provided by the client, if they are able to provide such
+a chain (see {{signature-algorithms}}).
+If the server cannot produce a certificate chain that is signed only via the
+indicated supported pairs, then it SHOULD continue the handshake by sending
+the client a certificate chain of its choice that may include algorithms
+that are not known to be supported by the client. This fallback chain MAY
+use the deprecated SHA-1 hash algorithm.
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: ... if the client indicated that it supported SHA-1 in combination with the signature algorithm in "signature_algorithms".

File: draft-ietf-tls-tls13.md (hunk 7 above)
Comment (ekr) [inline draft-ietf-tls-tls13.md]: Still an open issue.

File: draft-ietf-tls-tls13.md (hunk 8 above)
Comment (ekr) [inline draft-ietf-tls-tls13.md]: I am trying to decide whether it should go here or in CertificateVerify.

File: draft-ietf-tls-tls13.md (hunk 9 above)
Comment (ekr) [inline draft-ietf-tls-tls13.md]: Actually, I'm not sure that this is right, since we already agreed you can just send one that violates the rules.

File: draft-ietf-tls-tls13.md [hunk 10]
```diff
@@ -3107,101 +3267,45 @@ Structure of this message:
...
+hash.
 
 finished_label
 : For Finished messages sent by the client, the string
   "client finished".  For Finished messages sent by the server,
-  the string "server finished".
-{:br }
+  the string "server finished" [[OPEN ISSUE: Do we still need these?
+  labels? Should we expand them to indicate which phase we are
+  sending the messages in?]]
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: If you do keep the labels, then something needs to change.  Having both "server finished" and "server_finished" is going to be quite confusing.

File: draft-ietf-tls-tls13.md [hunk 11]
```diff
@@ -3381,50 +3495,36 @@ The following table describes the inputs to the key calculation for
...
-messages from the handshake where the configuration was established (including the
-type and length fields). Note that
-this requires the client and server to memorize these values.
-{:br }
+The handshake hash is defined as the hash of all handshake messages
+sent or received, starting at ClientHello up to the present time,
+with the exception of the Finished message, including the type and
+length fields of the handshake messages. This is the concatenation
+of all the exchanged Handshake structures in plaintext form (even if
+they were encrypted on the wire).
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: Need to explicitly exclude post-handshake messages. NewSessionTicket, CertificateRequest, etc...

File: draft-ietf-tls-tls13.md (hunk 8 above)
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: Ack, I don't think that it matters, but CertificateVerify has a certain merit to it.  Of course, we should merge all the messages and then it will matter much less.

File: draft-ietf-tls-tls13.md (hunk 9 above)
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: I think that this is important though.  If we allow inclusion of sha-1 in signature_algorithms, and not permit violations of the rules, then we have a way to signal that a client doesn't want sha-1 any more.  If you let servers just use sha-1 without signaling, then you are in a perpetual sha-1 limbo.

Comment (ekr) [inline draft-ietf-tls-tls13.md]: This is in the existing text, so why don't you send a PR. I suspect
@davegarrett would approve :)

File: draft-ietf-tls-tls13.md (hunk 11 above)
Comment (ekr) [inline draft-ietf-tls-tls13.md]: Also the 0-RTT messages.

**Comment by ekr** [comment]
I've rebased this version and made some updates. I need to go over the diff again before I merge so there may be some issues, but I wanted to put it up so others could comment ahead of that. Expected merge tonight or tomorrow.
